pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -157,6 +157,13 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 		bestHint = extendedHint
 	}
 
+	// the best hint might violate the NUMA allocation rule on which
+	// NUMA node cannot have both single and cross NUMA node allocations
+	// https://kubernetes.io/blog/2021/08/11/kubernetes-1-22-feature-memory-manager-moves-to-beta/#single-vs-cross-numa-node-allocation
+	if isAffinityViolatingNUMAAllocations(machineState, bestHint.NUMANodeAffinity) {
+		return fmt.Errorf("[memorymanager] preferred hint violates NUMA node allocation")
```

Comment:
where will it fail at otherwise? I am a bit surprised we need re-validation in the end instead of some fix in the way those hints calculated. But this is a comment from the very high level, I didn't dig into the code

Reply:
@SergeyKanzhelev Thank you for question.

> where will it fail at otherwise - in the example from bug description if CPUs needed were less than 3?

In the example above, if CPUs needed are less than 3, it will rejected at:
https://github.com/kubernetes/kubernetes/blob/ddd60de3f3aa21f4e0ffeab65049772e39cbccd8/pkg/kubelet/cm/memorymanager/policy_static.go#L515 
Because in the example above the condition ^ will be evaluated `true` for `NUMA0`. 
So it leaves us with `NUMA1` which is not enough to satisfy the memory request of `pod2`

Reply:
I think I understand it better now. But I am still not super deep in this code. @klueska do you have time for this one?